File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-3.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left = 0
        right = 0
        n = len(s)
        maxlen = 0
        currlen = 0
        window = {}
        while right < n:
            if s[right] not in window or window[s[right]] < left:
                currlen += 1
            else:
                left = window[s[right]] + 1
                maxlen = max(maxlen, currlen)
                currlen = right - left + 1
            window[s[right]] = right
            right += 1
        maxlen = max(maxlen, currlen)
        return maxlen

File: Data Structures & Algorithms/max-water-container/submission-3.java
class Solution {
    public int maxArea(int[] heights) {
        int left = 0;
        int right = heights.length - 1;
        int maxi = 0;
        while (left < right) {
            int width = right - left;
            int area = Math.min(heights[left], heights[right]) * width;
            maxi = Math.max(maxi, area);
            if (heights[left] <= heights[right]) {
                left++;
            } else {
                right--;
            }
        }
        return maxi;
    }

}